server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1109,6 +1118,30 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
+    private static List<SnapshotId> matchingSnapshotIds(RepositoryData repositoryData, Collection<String> snapshotsOrPatterns,
+                                                        String repositoryName) {
+        final Map<String, SnapshotId> allSnapshotIds = repositoryData.getSnapshotIds().stream().collect(
+                Collectors.toMap(SnapshotId::getName, Function.identity()));
+        final List<SnapshotId> foundSnapshots = new ArrayList<>();
+        for (String snapshotOrPattern : snapshotsOrPatterns) {
+            if (Regex.isSimpleMatchPattern(snapshotOrPattern) == false) {
+                final SnapshotId foundId = allSnapshotIds.get(snapshotOrPattern);
+                if (foundId == null) {
+                    throw new SnapshotMissingException(repositoryName, snapshotOrPattern);
+                } else {
+                    foundSnapshots.add(allSnapshotIds.get(snapshotOrPattern));
+                }
+            } else {
+                for (Map.Entry<String, SnapshotId> entry : allSnapshotIds.entrySet()) {
+                    if (Regex.simpleMatch(snapshotOrPattern, entry.getKey())) {
+                        foundSnapshots.add(entry.getValue());
```

Comment:
do we need a set? Isn't there a risk that we're adding the same item multiple times if there are overlapping wildcards?

Reply:
Right ... my bad fixed + added an assertion and test that would've tripped it to cover this.